clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2100,223 +2141,164 @@ public DescribeConsumerGroupsResult describeConsumerGroups(final Collection<Stri
+                                    new MemberDescription(
+                                        groupMember.memberId(),
+                                        groupMember.clientId(),
+                                        groupMember.clientHost(),
+                                        new MemberAssignment(assignment.partitions()));
+                                consumers.add(memberDescription);
                             }
+                            final String protocol = groupMetadata.protocol();
+                            final ConsumerGroupDescription consumerGroupDescription =
+                                new ConsumerGroupDescription(groupId, protocolType.isEmpty(), consumers, protocol);
+                            future.complete(consumerGroupDescription);
                         }
-
-                        @Override
-                        void handleFailure(Throwable throwable) {
-                            completeAllExceptionally(consumerGroupFutures.values(), throwable);
-                        }
-                    }, nowDescribeConsumerGroups);
-
-                    resultFutures.complete(new HashMap<String, KafkaFuture<ConsumerGroupDescription>>(consumerGroupFutures));
+                    }
                 }
 
                 @Override
                 void handleFailure(Throwable throwable) {
-                    resultFutures.completeExceptionally(throwable);
+                    completeAllExceptionally(consumerGroupFutures.values(), throwable);
                 }
-            }, nowFindCoordinator);
+            }, nowDescribeConsumerGroups);
+
         }
 
-        return new DescribeConsumerGroupsResult(resultFutures);
+        return new DescribeConsumerGroupsResult(new HashMap<String, KafkaFuture<ConsumerGroupDescription>>(consumerGroupFutures));
     }
 
     @Override
     public ListConsumerGroupsResult listConsumerGroups(ListConsumerGroupsOptions options) {
-        final KafkaFutureImpl<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> nodeAndConsumerGroupListing = new KafkaFutureImpl<>();
+        final Collection<Node> nodes = listNodes(options.timeoutMs());
 
-        final long nowMetadata = time.milliseconds();
-        final long deadline = calcDeadlineMs(nowMetadata, options.timeoutMs());
-
-        runnable.call(new Call("listNodes", deadline, new LeastLoadedNodeProvider()) {
-            @Override
-            AbstractRequest.Builder createRequest(int timeoutMs) {
-                return new MetadataRequest.Builder(Collections.<String>emptyList(), true);
-            }
-
-            @Override
-            void handleResponse(AbstractResponse abstractResponse) {
-                MetadataResponse metadataResponse = (MetadataResponse) abstractResponse;
-
-                final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashMap<>();
+        final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashMap<>();
```

Comment:
We can still follow the original pattern of calling a `listNodes` to any broker, get the result, and inside the response handling logic based on the result sending a `listConsumerGroups` to each node, and then group and flatten the result so we hide the node information of each sub-map it is belonging to.